### DISCUSSION TITLE
Conditional "use"

### DISCUSSION THREAD

**Discussion by ion1** [gleam-lang/gleam#2824] [Ideas & suggestions]
# The proposal

It might be nice to have syntactic sugar for a conditional `use <- something`, for example `use <- iterator.yield(...)`.

This video demonstrates the desugaring for a proposed syntax in which the final `use <- foo(args)` within each branch of the case expression becomes `foo(args, next)` where `next` stands for the code following the case expression as a function.

https://github.com/gleam-lang/gleam/assets/39660/6c18faff-c1d9-4e27-93a6-f82493783923

`noop` is just a function such that `noop(next)` becomes `next()` in order not to do yield anything within that branch.

```gleam
fn noop(next: fn() -> a) -> a {
  next()
}
```

## Alternatives

As I realized during the following discussion, the same can be achieved in the current language as follows:

```gleam
  use <- iterator.yield("foo")
  use <- case add_bar {
    True -> fn(next) {
      use <- iterator.yield("bar 1")
      iterator.yield("bar 2", next)
    }
    False -> noop
  }
  use <- iterator.yield("baz")
  iterator.empty()
```

Or in the case of a single `iterator.yield` in a branch, not needing a block:

```gleam
  use <- iterator.yield("foo")
  use <- case add_bar {
    True -> iterator.yield("bar", _)
    False -> noop
  }
  use <- iterator.yield("baz")
  iterator.empty()
```

It is not too bad without the additional syntactic sugar, just a bit asymmetric with the explicit "next" parameter.

## `noop`

Whether the syntactic sugar is added or not, the `noop` function might be useful for anyone utilizing a conditional `use`. Should it perhaps be in the standard library?

## Original message

EDIT: Updated the message to hopefully be clearer, as the original message resulted in confusion in the following discussion.

<details>
  <summary>The original message</summary>

Is there a nice way to do a conditional `use <- something`, for example `use <- iterator.yield(...)`? I'm new to the language and I didn't get suggestions when I asked on the Discord server.

In case there isn't, my proposal follows.

I would like to be able to write code such as
```gleam
use <- iterator.yield("foo")
case add_bar {
  True -> use <- iterator.yield("bar")
  False -> use()
}
use <- iterator.yield("baz")
iterator.empty()
```
where the `use()` stands in for whatever syntax would be adopted for a no-op "use".

My proposal is the syntactic sugar where
```gleam
case VALUE {
  PATTERN1 -> use <- fn1(...)
  PATTERN2 -> use <- fn2(...)
  PATTERN3 -> {
    // ...
    use <- fn3(...)
  }
  PATTERN4 -> use()
}
```
would desugar into
```gleam
use <- fn(cont) {
  case VALUE {
    PATTERN1 -> fn1(..., cont)
    PATTERN2 -> fn2(..., cont)
    PATTERN3 -> {
      // ...
      fn3(..., cont)
    }
    PATTERN4 -> cont()
  }
}
```

Given a result value,
```gleam
let result = case VALUE {
  PATTERN1 -> use <- fn1(...) // Note: not "use result <- fn1(...)"
  PATTERN2 -> use(result_value)
}
```
would desugar into
```gleam
use result <- fn(cont) {
  case VALUE {
    PATTERN1 -> fn1(..., cont)
    PATTERN2 -> cont(result_value)
  }
}
```
</details>

**Comment by lpil** [comment]
Use doesn't have any particular semantics, it's just sugar for a function call. Given the function could do anything how would we know what the no-op could be?

What would your example look like rewritten using the regular function call syntax instead of use?

**Reply by ion1** [reply]
```gleam
iterator.yield("foo", fn() {
  fn(cont) {
    case add_bar {
      True -> iterator.yield("bar", cont)
      False -> cont()
    }
  }(fn() {
    iterator.yield("baz", fn() {
      iterator.empty()
    })
  })
})
```

The no-op "use" would only have a meaning within case expressions if they were to be desugared into the `use <- fn(cont) { }` form, I suppose.

The no-op "use" would be equivalent to:
```gleam
use <- fn(cont) { cont() }
```
or in the case of a result value:
```gleam
use result <- fn(cont) { cont(value) }
```

**Reply by lpil** [reply]
How would the compiler know what function to use as cont? It seems which one you want would be different depending on what function you "use"

**Reply by ion1** [reply]
`use <- fn(cont) { cont() }` works in Gleam right now. `cont` will be the rest of the code following the use expression.

In my proposed desugaring for use-within-case, `cont` will be the rest of the code following the case expression.

Let's forget the no-op "use" for the moment, as the main point is the conditional "use". For now, let's use
```gleam
fn noop(cont: fn() -> a) -> a {
  cont()
}
```
instead.

If you desugar
```gleam
case True {
  True -> use <- iterator.yield("hello")
  False -> use <- noop()
}
iterator.empty()
```
into
```gleam
use <- fn(cont) {
  case True {
    True -> iterator.yield("hello", cont)
    False -> noop(cont)
  }
}
iterator.empty()
```
and (using the existing desugaring rules) further into
```gleam
fn(cont) {
  case True {
    True -> iterator.yield("hello", cont)
    False -> noop(cont)
  }
}(fn() {
  iterator.empty()
})
```
there is no ambiguity about what `cont` is at any point.

**Reply by lpil** [reply]
Sorry, I don't understand. Can you show what this would desugar to (so without "use" at all).

**Reply by ion1** [reply]
Sorry for failing to explain clearly enough and thanks for your patience! Here is a full program in which `example1` demonstrates the desugaring of `example0` I'm proposing, and `example2` demonstrates how Gleam desugars `example1` right now AFAIU. `example2` is without "use" at all.

```gleam
import gleam/iterator.{type Iterator}

pub fn main() {
  // let assert [] = example0(False) |> iterator.to_list()
  let assert [] = example1(False) |> iterator.to_list()
  let assert [] = example2(False) |> iterator.to_list()

  // let assert ["hello"] = example0(True) |> iterator.to_list()
  let assert ["hello"] = example1(True) |> iterator.to_list()
  let assert ["hello"] = example2(True) |> iterator.to_list()

  let assert "foo" = {
    use <- noop()
    "foo"
  }
}

// Proposed syntax:
// fn example0(add_hello: Bool) -> Iterator(String) {
//   case add_hello {
//     True -> use <- iterator.yield("hello")
//     False -> use <- noop()
//   }
//   iterator.empty()
// }

// After desugaring the case expression:
fn example1(add_hello: Bool) -> Iterator(String) {
  use <- fn(cont) {
    case add_hello {
      True -> iterator.yield("hello", cont)
      False -> noop(cont)
    }
  }
  iterator.empty()
}

// After desugaring the remaining use expression:
fn example2(add_hello: Bool) -> Iterator(String) {
  fn(cont) {
    case add_hello {
      True -> iterator.yield("hello", cont)
      False -> noop(cont)
    }
  }(fn() {
    iterator.empty()
  })
}

fn noop(cont: fn() -> a) -> a {
  cont()
}
```

**Comment by ion1** [comment]
I realized that it can be achieved this way without the syntactic sugar for a non-block expression:
```gleam
  use <- iterator.yield("foo")
  use <- case add_bar {
    True -> iterator.yield("bar", _)
    False -> noop
  }
  use <- iterator.yield("baz")
  iterator.empty()
```
given the function
```gleam
fn noop(cont: fn() -> a) -> a {
  cont()
}
```

One drawback I can think of is that the mandatory `_` capture makes it look a bit different from the unconditional yields.

Another is that doing two yields in the branch requires something like:
```gleam
  use <- iterator.yield("foo")
  use <- case add_bar {
    True -> fn(cont) {
      use <- iterator.yield("bar 1")
      iterator.yield("bar 2", cont)
    }
    False -> noop
  }
  use <- iterator.yield("baz")
  iterator.empty()
```
while with the proposed syntactic sugar it would look more consistent:
```gleam
  use <- iterator.yield("foo")
  case add_bar {
    True -> {
      use <- iterator.yield("bar 1")
      use <- iterator.yield("bar 2")
    }
    False -> use <- noop()
  }
  use <- iterator.yield("baz")
  iterator.empty()
```

**Reply by lpil** [reply]
That first version is a lot clearer to me, I'm still not quite getting how the proposal works. Given `use` is the most complex and confusing thing in Gleam I think we wanna keep it as simpler as possible, and we generally only want to add new features that enable new things you could not write previously in Gleam.

**Reply by ion1** [reply]
> I think we wanna keep it as simpler as possible, and we generally only want to add new features that enable new things you could not write previously in Gleam.

Fair enough. However, hoping to at least manage to explain what I had in mind, here's one more attempt at doing so:

https://github.com/gleam-lang/gleam/assets/39660/6c18faff-c1d9-4e27-93a6-f82493783923
